src/GraphQL/DI/GraphQLBuilderBase.cs

```diff
@@ -0,0 +1,76 @@
+using System;
+using GraphQL.Caching;
+using GraphQL.Execution;
+using GraphQL.Types;
+using GraphQL.Validation;
+using GraphQL.Validation.Complexity;
+
+namespace GraphQL.DI
+{
+    /// <summary>
+    /// Base implementation of <see cref="IGraphQLBuilder"/>.
+    /// </summary>
+    public abstract class GraphQLBuilderBase : IGraphQLBuilder
+    {
+        /// <summary>
+        /// Register the default services required by GraphQL if they have not already been registered.
+        /// <br/><br/>
+        /// Does not include <see cref="IDocumentWriter"/>, and the default <see cref="IDocumentExecuter"/>
+        /// implementation does not support subscriptions.
+        /// </summary>
+        protected virtual void Initialize()
```

Comment:
👍 for all `TryRegister` here. The thought is correct, but it seems to me that it is implemented with a disadvantage - actually all these registrations will be called since user calls `AddGraphQL` method first (and creates `GraphQLBuilder`) and only then calls fluent continuations. At least I am sure that the majority do that.
Consider such API instead:
```csharp
    public static IGraphQLBuilder AddGraphQL(this IServiceCollection services, Action<IGraphQLBuilder> configure)
        {
            var builder = new GraphQLBuilder(services);
            configure(builder);
            builder.Initialize(); // May be renamed into TryRegisterDefaults() or RegisterDefaultServices()
            return builder;
        }
```

The idea is to locate GraphQL specifics ***inside* `AddGraphQL` method. The advantage is that the user configuration will always have priority. At the same time, nothing prevents the configuration outside the method.

I prefer this approach with the placement of fluent configuration within the action delegate. I like that the specific functionality (all GraphQL bits in our case) is grouped into `AddGraphQL` and not scattered in composition root.

Reply:
If we have an example of this design pattern in use by Microsoft then I would consider it. I’d like to do whatever is common, or what is one of the generally accepted patterns.

Reply:
I see your point but I don’t think the current pattern is “bad”, just different. I’m fine with changing it so long as, like I said, it’s a common and accepted design pattern.

Reply:
One benefit is that we could perhaps eliminate the replace argument. Or make it unnecessary.

Reply:
Examples:
```
public static IServiceCollection AddRouting(this IServiceCollection services, Action<RouteOptions> configureOptions)
public static IApplicationBuilder UseEndpoints(this IApplicationBuilder builder, Action<IEndpointRouteBuilder> configure)
```

This is the first thing that comes to mind.

> the current pattern is “bad”, just different.

Of course it's not bad. Both methods work. The difference can be almost invisible, but it exists.

Reply:
I may post a PR after reviewing this adn some other stuff.

Reply:
I would suggest that we provide a configuration delegate overload with the service provider (as well as one without)

Reply:
Seems so.

Reply:
> I would suggest that we provide a configuration delegate overload with the service provider (as well as one without)

Nevermind, I wasn't thinking.  This is of course impossible, since the builder is building the service collection.

Reply:
#2782